modules/oe_content_sub_entity/src/SubEntitySubscriberBase.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\oe_content_sub_entity;
+
+use Drupal\Core\Entity\ContentEntityInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Drupal\oe_content_sub_entity\Event\SubEntityEvents;
+use Drupal\oe_content_sub_entity\Event\SubEntityLabelInformationEvent;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Provides a base implementation for SubEntitySubscriber.
+ */
+abstract class SubEntitySubscriberBase implements EventSubscriberInterface {
+
+  use StringTranslationTrait;
+
+  /**
+   * Checking if particular even subscriber applicable for specific entity.
+   *
+   * @param \Drupal\Core\Entity\ContentEntityInterface $entity
+   *   The content entity.
+   *
+   * @return bool
+   *   TRUE if this even subscriber should be used or FALSE to let other event
+   *   subscriber decide.
+   */
+  abstract protected function applies(ContentEntityInterface $entity): bool;
+
+  /**
+   * Form labels for specific sub-entity types or bundles.
+   *
+   * @param \Drupal\Core\Entity\ContentEntityInterface $entity
+   *   The content entity.
+   *
+   * @return string|\Drupal\Core\StringTranslation\TranslatableMarkup|null
+   *   The generated label.
+   */
+  abstract protected function generateLabel(ContentEntityInterface $entity);
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents() {
+    return [
+      SubEntityEvents::LABEL_FORMATION => ['onLabelFormation'],
+    ];
+  }
+
+  /**
+   * Extracting sub-entity label which depends on entity bundle.
+   *
+   * @param \Drupal\oe_content_sub_entity\Event\SubEntityLabelInformationEvent $event
+   *   Sub-entity event.
+   */
+  public function onLabelFormation(SubEntityLabelInformationEvent $event): void {
+    if ($this->applies($event->getEntity())) {
+      $generated_label = $this->generateLabel($event->getEntity());
+      if (!empty($generated_label)) {
+        $event->setLabel($generated_label);
+      }
+    }
+  }
+
+  /**
+   * Default label generator for target sub-entity.
+   *
+   * @param \Drupal\Core\Entity\ContentEntityInterface $entity
+   *   The content entity.
+   *
+   * @return string
+   *   Generated entity label.
+   */
+  protected function defaultLabel(ContentEntityInterface $entity): ?string {
+    $labels = $this->getReferencedEntityLabels($entity);
+    if (!empty($labels)) {
+      return $labels;
+    }
+    return NULL;
+  }
+
+  /**
+   * Gets labels of referenced entities.
+   *
+   * @return string
+   *   Labels separated by comma.
+   */
+  protected function getReferencedEntityLabels(ContentEntityInterface $entity): string {
+    // Load referenced entities.
+    $entities = $entity->referencedEntities();
+
+    $labels = [];
+    foreach ($entities as $entity) {
+      if ($entity instanceof ContentEntityInterface && $entity->getEntityType()->hasKey('label')) {
+        $labels[] = $entity->label();
```

Comment:
I assume that it doesn't provide translated labels that can be important for Author sub entity. Please see https://github.com/openeuropa/oe_content/pull/463/files#diff-9347041f0d1e9f4bf0f750dca978f2990c0589dbff59feb376ae15465da8ce88R25. I can't test it because this branch should be rebased since main EWPP-1410 is updated.

Reply:
I don't sure if it should work according to requirements where the reference field to Person bundle is declared as not translatable.

Reply:
Relation to the Person is untranslatable, but Person node is translatable, so we need to show french label from Person node on the french page.
<img width="1104" alt="Screenshot 2021-09-22 at 18 52 47" src="https://user-images.githubusercontent.com/1183951/134378332-187b8dc7-c7a1-404d-b083-fb0a82aed814.png">

This example from the form, but if we will need to show label on the page it will be untranslatable

Reply:
It was a suggestion. We can't use `\Drupal::service('entity.repository')` here. Container injection for the event you had to implement yourself...